services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java

```diff
@@ -196,43 +195,43 @@ public Count getOpenInvitationCountForUser(String principalId) {
-		List<String> emails = new ArrayList<>();
-		for (PrincipalAlias alias : aliases) {
-			if (alias.getType() == AliasType.USER_EMAIL) {
-				emails.add(alias.getAlias());
-			}
+		MembershipInvtnSubmission mis = membershipInvtnSubmissionDAO.get(membershipInvitationId);
+		if (mis.getExpiresOn() != null && mis.getExpiresOn().before(new Date())) {
+			throw new IllegalArgumentException("Indicated MembershipInvtnSubmission is expired.");
+		}
+		if (!(mis.getInviteeId() == null && mis.getInviteeEmail() != null)) {
+			throw new IllegalArgumentException("The indicated invitation must have a null inviteeId and a non null inviteeEmail.");
 		}
-		// Get inviteeEmail from membershipInvitation
-		String inviteeEmail = membershipInvtnSubmissionDAO.getInviteeEmail(membershipInvitationId);
-		// If inviteeEmail is in the emails list, construct InviteeVerificationSignedToken and return it
-		if (emails.contains(inviteeEmail)) {
-			InviteeVerificationSignedToken token = new InviteeVerificationSignedToken();
-			token.setInviteeId(userId.toString());
-			token.setMembershipInvitationId(membershipInvitationId);
-			SignedTokenUtil.signToken(token);
-			return token;
+		if (!principalAliasDAO.aliasIsBoundToPrincipal(mis.getInviteeEmail(), Long.toString(userId))) {
+			// inviteeEmail is not associated with the user
+			throw new UnauthorizedException("This membership invitation is not addressed to the authenticated user.");
 		}
-		// inviteeEmail is not associated with the user, return null to convey invitee verification failure
-		return null;
+		InviteeVerificationSignedToken response = new InviteeVerificationSignedToken();
+		response.setInviteeId(userId.toString());
+		response.setMembershipInvitationId(membershipInvitationId);
+		response.setExpiresOn(new Date(new Date().getTime() + TWENTY_FOUR_HOURS_IN_MS));
+		SignedTokenUtil.signToken(response);
+		return response;
 	}
 
 	@WriteTransactionReadCommitted
 	@Override
 	public void updateInviteeId(Long userId, String misId, InviteeVerificationSignedToken token) {
-		AuthorizationStatus status = authorizationManager.canAccessMembershipInvitationSubmission(userId, misId, token, ACCESS_TYPE.UPDATE);
+		AuthorizationStatus status = authorizationManager.canAccessMembershipInvitationSubmission(userId, token, ACCESS_TYPE.UPDATE);
 		if (!status.getAuthorized()) {
 			throw new UnauthorizedException(status.getReason());
 		}
-		MembershipInvtnSubmission mis = membershipInvtnSubmissionDAO.get(misId);
-		if (!(mis.getInviteeId() == null && mis.getInviteeEmail() != null)) {
-			throw new UnauthorizedException("Cannot update inviteeId. The target invitation must have a null inviteeId and a non null inviteeEmail.");
+		if (!misId.equals(token.getMembershipInvitationId())) {
+			throw new IllegalArgumentException("ID in URI and ID in signed token don't match");
+		}
+		if (token.getExpiresOn() != null && token.getExpiresOn().before(new Date())) {
```

Comment:
if (token.getExpiresOn())==null) throw new IAE("expiration is missing.");
if (token.getExpiresOn().before(new Date())) {) throw new IllegalArgumentException("InviteeVerificationSignedToken is expired");